### PR TITLE
Static authorization server interceptor implementation

### PR DESCRIPTION
**Pull request by ashithasantosh**
1. Implements static authorization server interceptor with end to end tests.
2. Adds support for "v3" HeaderMatcher proto by handling new fields like string_match.

### CODE REVIEW

**Review by ejona86 — COMMENTED** [review]
Was there a reason you didn't use RbacFilter directly?

CC @YifeiZhuang

**Comment by YifeiZhuang** [comment]
> > Was there a reason you didn't use RbacFilter directly?
> 
> I moved the rbac parsing logic to a common location in io.grpc.xds.internal.rbac.engine package due to following reasons
> 
> 1. to avoid dependency on xds. Also RbacFilter is not public in io.grpc.xds; cannot be accessed from outside package
> 2. for code readability. the code seemed more readable to me if we split the rbac parsing code into a different class, instead of depending on RbacFilter implementation


I think Eric is saying you only need to call [`rbacfilter.buildServerIntercetor()` ](https://github.com/grpc/grpc-java/blob/e939bf6fb89a52e2afc7f830a787f73e58af3807/xds/src/main/java/io/grpc/xds/RbacFilter.java#L171-L193)
but yea we have to make `RbacFilter` accessible from auth

**Comment by ashithasantosh** [comment]
> I think Eric is saying you only need to call [`rbacfilter.buildServerIntercetor()` ](https://github.com/grpc/grpc-java/blob/e939bf6fb89a52e2afc7f830a787f73e58af3807/xds/src/main/java/io/grpc/xds/RbacFilter.java#L171-L193) but yea we have to make `RbacFilter` accessible from auth

In non-xDS authorization, we have upto two RBAC engines per Server Interceptor. And the grpc user can use the factory method to create the interceptor and attach it to server enabling grpc authorization.
https://docs.google.com/document/d/1SGE1K38wYtSHDdJm_0sip9F2ozGczqepsxKRANbw6-c/edit?usp=sharing

If we use rbac filter's method to generate server interceptor, then we would have upto two interceptors. This would complicate the API that we expose to the users. Also, when we introduce dynamic file reloading support, that is we reload the authorization policy periodically from file system, we would keep on creating new server interceptors as the policy changes which would really complicate the design, instead with the current design we will just update the internal RBAC engines in the server interceptor.

**Comment by ashithasantosh** [comment]
> > This would complicate the API that we expose to the users
> 
> No. Two vs 1 interceptor internally should have no bearing on the API we expose to users.
> 
Right! I meant the same thing above (quoting previous message "or we could have two interceptors within an interceptor and keep the API unchanged")
I updated the implementation based on the comment. Please take a look. Sorry it took so long. I was on a month long vacation and was busy catching up on some work.

**Comment by ashithasantosh** [comment]
@ejona86 Please take a look at the PR when you find time. Thank you:)

**Comment by ashithasantosh** [comment]
@ejona86 Friendly ping!:)

**Comment by ashithasantosh** [comment]
@ejona86 PTAL

File: authz/src/main/java/io/grpc/authz/AuthorizationServerInterceptor.java
Comment (ejona86) [inline authz/src/main/java/io/grpc/authz/AuthorizationServerInterceptor.java:37]: `@ExperimentalApi`

File: authz/src/main/java/io/grpc/authz/AuthorizationServerInterceptor.java
Comment (ejona86) [inline authz/src/main/java/io/grpc/authz/AuthorizationServerInterceptor.java:47]: We don't want to expose the types involved here as-is. We try to hide them. I think the easiest thing to do is create a `io.grpc.xds.InternalRbacFilter` public class with `@Internal` annotation. Give it a `static` method akin to `ServerInterceptor create(RBAC config)` and have it have this logic for throwing on error. Basically, that then means we only have to expose a single method from a single internal class.

**Comment by ejona86** [comment]
@YifeiZhuang, do you know how we were missing contains and stringMatcher?

**Comment by ejona86** [comment]
> @YifeiZhuang, do you know how we were missing contains and stringMatcher?

Oh, apparently they aren't used by this code, so that resolves some of my questions. I think they were just added later. @ashithasantosh simply noticed they were missing when comparing to C++, which makes sense.

File: authz/src/main/java/io/grpc/authz/AuthorizationServerInterceptor.java
Comment (ashithasantosh) [reply authz/src/main/java/io/grpc/authz/AuthorizationServerInterceptor.java:37]: Done.

File: authz/src/main/java/io/grpc/authz/AuthorizationServerInterceptor.java
Comment (ashithasantosh) [reply authz/src/main/java/io/grpc/authz/AuthorizationServerInterceptor.java]: Done.

**Review by YifeiZhuang — APPROVED** [review]

File: authz/src/test/java/io/grpc/authz/AuthorizationEnd2EndTest.java
Comment (ashithasantosh) [reply authz/src/test/java/io/grpc/authz/AuthorizationEnd2EndTest.java:46]: Updated the tests. PTAL

File: xds/src/main/java/io/grpc/xds/internal/Matchers.java
Comment (ashithasantosh) [reply xds/src/main/java/io/grpc/xds/internal/Matchers.java]: Done.

**Review by ejona86 — COMMENTED** [review]
This looks good to me (modulo my comments), for the parts I've looked at. I'll let Yifei do the final full review.

File: xds/src/main/java/io/grpc/xds/InternalRbacFilter.java
Comment (ashithasantosh) [reply xds/src/main/java/io/grpc/xds/InternalRbacFilter.java:28]: Done

File: authz/src/main/java/io/grpc/authz/AuthorizationServerInterceptor.java
Comment (ashithasantosh) [reply authz/src/main/java/io/grpc/authz/AuthorizationServerInterceptor.java:70]: Done.

File: authz/src/main/java/io/grpc/authz/AuthorizationServerInterceptor.java
Comment (ashithasantosh) [reply authz/src/main/java/io/grpc/authz/AuthorizationServerInterceptor.java]: Oh, I didn't know. Thank you for the information. Removed the exception.

**Comment by ashithasantosh** [comment]
@ejona86 Thank you for the review. PTAL I resolved all the comments. 
@YifeiZhuang already approved the PR, should I wait on a final review?

File: authz/src/main/java/io/grpc/authz/AuthorizationServerInterceptor.java
Comment (ejona86) [inline authz/src/main/java/io/grpc/authz/AuthorizationServerInterceptor.java]: Don't mention GrpcAuthorizationEngine; that is an implementation detail. "to Envoy RBAC policies to make authorization decisions."

File: authz/src/main/java/io/grpc/authz/AuthorizationServerInterceptor.java
Comment (ejona86) [inline authz/src/main/java/io/grpc/authz/AuthorizationServerInterceptor.java]: This is also a detail and over-specifies. We could probably just remove this sentence and be okay.

File: authz/src/main/java/io/grpc/authz/AuthorizationServerInterceptor.java
Comment (ejona86) [inline authz/src/main/java/io/grpc/authz/AuthorizationServerInterceptor.java]: "This provider" is weird, because there's no provider. And it is written from the internal perspective. How about, "The policy cannot be changed once created." (Then when we have the updating interceptor we could have a "To change the policy after creation, see TheOtherClass".)

File: authz/src/main/java/io/grpc/authz/AuthorizationServerInterceptor.java
Comment (ejona86) [inline authz/src/main/java/io/grpc/authz/AuthorizationServerInterceptor.java:35]: Have "gRPC Authorization policy" be a link to https://github.com/grpc/proposal/blob/master/A43-grpc-authorization-api.md#user-facing-authorization-policy

File: xds/src/main/java/io/grpc/xds/InternalRbacFilter.java
Comment (ejona86) [reply xds/src/main/java/io/grpc/xds/InternalRbacFilter.java:28]: Make it private. `protected` isn't bad, but `private` is ideal.

File: authz/src/main/java/io/grpc/authz/AuthorizationServerInterceptor.java
Comment (ashithasantosh) [reply authz/src/main/java/io/grpc/authz/AuthorizationServerInterceptor.java]: Done

File: authz/src/main/java/io/grpc/authz/AuthorizationServerInterceptor.java
Comment (ashithasantosh) [reply authz/src/main/java/io/grpc/authz/AuthorizationServerInterceptor.java]: Agreed! I had removed "provider" in the last commit.
https://github.com/grpc/grpc-java/pull/8934/commits/fb79c7472d88407f6b4057558678be0b57d98e67
Updated the comment. Thank you for the suggestion:)

File: authz/src/main/java/io/grpc/authz/AuthorizationServerInterceptor.java
Comment (ashithasantosh) [reply authz/src/main/java/io/grpc/authz/AuthorizationServerInterceptor.java]: Removed the line

File: authz/src/main/java/io/grpc/authz/AuthorizationServerInterceptor.java
Comment (ashithasantosh) [reply authz/src/main/java/io/grpc/authz/AuthorizationServerInterceptor.java]: Sounds good. Updated comment.

**Review by ejona86 — APPROVED** [review]

**Pull request merged**
